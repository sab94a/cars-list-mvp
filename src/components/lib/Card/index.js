// @flow

import React from 'react';
import cx from 'classnames';
import type { Card as Props} from 'components/lib/types';
import styles from './index.module.scss';

const Card = ({ image, title, description, footer, loading, elemType = 'div' }: Props) => {
    const rootClasses = cx(styles.root, { [styles.loading]: loading });
    const imageStyle = { backgroundImage: loading || !image ? null : `url(${ image })` }
    const Component = elemType
    return (
        <Component className={ rootClasses }>
            <div className={ styles.image } style={ imageStyle } />
            <div className={ styles.content }>
                <h3 className={ styles.title }>{ title }</h3>
                <p className={ styles.description }>{ description }</p>
                <div className={ styles.footer }>
                    { footer }
                </div>
            </div>
        </Component>
    );
};

export default React.memo<Props>(Card);
