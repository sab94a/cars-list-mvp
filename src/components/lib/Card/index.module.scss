@import 'components/lib/variables.scss';

.root {
    display: flex;
    width: 100%;
    padding: $offset-sm;
    border: $border-width-sm solid $light;
    margin-top:  $offset-md;

    &.loading {
        .image, .title, .description, .footer > * {
            color:  $light;
            background: $light;
        }

        .title {
            height: 26px;
        }

        .description {
            height: 14px;
        }

        .footer:after {
            content: '';
            @extend .footerLink;
            width: 65px;
            height: 14px;
            color:  $light;
            background: $light;
        }
    }
}

.image {
    width: 90px;
    background-position: center;
    background-size: contain;
    background-repeat: no-repeat;
}

.content {
    padding: 3px 0 3px $offset-lg;
    min-width: 245px;
}

.title {
    margin: 0;
    margin-bottom: 2px;
    padding-bottom: $offset-sm;
    line-height: 1;
    font-size: $font-size-md;
    font-weight: bold;
}

.description {
    margin: 0;
}

.footerLink {
    display: inline-block;
    vertical-align: middle;
    margin-right: $border-width;
    text-decoration: none;
    margin-top: $offset-sm;
}

.footer {
    & > * {
        @extend .footerLink;
    }
}
