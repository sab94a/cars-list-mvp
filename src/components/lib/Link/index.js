// @flow

import React from 'react';
import cx from 'classnames';
import type { AbstractComponent } from 'react';
import type { Link as Props } from 'components/lib/types';
import styles from './index.module.scss';

class Link extends React.Component<Props> {
    static Component: AbstractComponent<*> | string = 'a';

    render() {
        const { href, children, className, ...rest } = this.props;

        return (
            <Link.Component className={cx( className, styles.root )} { ...rest } href={ href }>{ children }</Link.Component>
        );
    }
}

export default Link;
