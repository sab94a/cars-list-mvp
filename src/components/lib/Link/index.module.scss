@import 'components/lib/variables.scss';

.root {
    color: $primary;

    &:hover, &:focus {
        text-decoration: underline;
        color: $primary-pressed;    
    }
}
