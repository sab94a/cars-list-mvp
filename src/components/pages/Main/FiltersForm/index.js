//@flow 

import React from 'react';
import Select from 'components/lib/Select';
import Button from 'components/lib/Button';
import styles from './index.module.scss';

export type Props = {
    color: string,
    manufacturer: string,
    colors: Array<string>,
    manufacturers: Array<string>
}

class FiltersForm extends React.PureComponent<Props> {
    static Titles = {
        Color: 'Color',
        Manufacturer: 'Manufacturer',
        Submit: 'Filter'
    }

    static Placeholders = {
        Color: 'All car colors',
        Manufacturer: 'All manufacturers'
    }

    static Fields = {
        Color: 'color',
        Manufacturer: 'manufacturer'
    }
    
    static defaultProps = {
        onSubmit: () => null
    }
    
    state = {
        color: this.props.color,
        manufacturer: this.props.manufacturer
    }

    handlers = {}

    bindChange(field) {
        if(!this.handlers[field]) {
            this.handlers[field] = value => {
                this.setState({
                    [field]: value
                })
            }
        }
        return this.handlers[field];
    };

    onSubmit = e => {
        e.preventDefault();

        this.props.onSubmit(this.state)
    };

    componentDidUpdate(prevProps) {
        // I know about this article https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#when-to-use-derived-state
        // But I think that in this expamle it's probably only one solution
        // Because we should have it's own state and we should update state, when props changed
        // For example if props changed from Url Query params

        const { color, manufacturer } = this.props;

        let state = {};

        if (color !== prevProps.color) {
            state.color = color
        };

        if (manufacturer !== prevProps.manufacturer) {
            state.manufacturer = manufacturer
        };

        if(Object.keys(state).length) {
            this.setState(state)
        };
    }

    render() {
        const { colors, manufacturers } = this.props;
        const { color, manufacturer } = this.state;

        return (
            <form className={ styles.root } onSubmit={ this.onSubmit }>
                <Select 
                    className={ styles.select } 
                    value={ color }
                    onChange={ this.bindChange(FiltersForm.Fields.Color) }
                    label={ FiltersForm.Titles.Color }
                    placeholder={ FiltersForm.Placeholders.Color }
                    options={ colors } 
                />
                <Select 
                    className={ styles.select } 
                    value={ manufacturer }
                    onChange={ this.bindChange(FiltersForm.Fields.Manufacturer) }
                    label={ FiltersForm.Titles.Manufacturer }
                    placeholder={ FiltersForm.Placeholders.Manufacturer }
                    options={ manufacturers } 
                />
                <Button type='submit'>{ FiltersForm.Titles.Submit }</Button>
            </form>
        )
    }
}

export default FiltersForm;
