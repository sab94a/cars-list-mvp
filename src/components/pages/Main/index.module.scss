@import 'components/lib/variables.scss';

.root {
    display: flex;
    padding: $offset-lg;
    padding-bottom: 0;
}

.aside {
    padding-right: $offset-lg;
}

.content {
    flex-grow: 1;
}

.searchHeader {
    display: flex;
    padding-bottom: $offset-lg;
    justify-content: space-between;

    h2 {
        margin: 0 0 $offset-md 0;
        line-height: 1;
    }
}
.searchResult {
    font-size: $font-size-md;
}

.items {
    margin: 0;
    padding: 0;

    li:first-of-type {
        margin-top: 0;
    }
}

.notFound {
    padding: $offset-md 0;
}
