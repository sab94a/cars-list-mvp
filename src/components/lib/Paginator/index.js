// @flow

import React from 'react';
import qs from 'query-string';
import cx from 'classnames';
import Link from 'components/lib/Link';
import type { Paginator as Props } from 'components/lib/types';
import styles from './index.module.scss';

class Paginator extends React.PureComponent<Props> {
    static Query = 'page'
    static Titles = {
        First: 'First',
        Prev: 'Previous',
        Next: 'Next',
        Last: 'Last'
    }

    buildQuerySearch(page: number):?string {
        const { total, active } = this.props;

        if (page === active || page < 1 || page > total) {
            return null;
        }

        let params = qs.parse(window.location.search);

        return `?${qs.stringify({
            ...params,
            [Paginator.Query]: page
        })}`
    }

    renderLink(link: ?string, title:string) {
        return link ? (
            <Link className={ styles.link } href={ link }>{ title }</Link>
        ) : (
            <span className={ cx(styles.link, styles.disable) }>{ title }</span>
        )
    }

    render() {
        const { active, total } = this.props;

        return (
            <div className={ styles.root }>
                { this.renderLink(this.buildQuerySearch(1), Paginator.Titles.First) }
                { this.renderLink(this.buildQuerySearch(active - 1), Paginator.Titles.Prev) }
                <div className={ styles.content }>{`Page ${ active } of ${ total }`}</div>
                { this.renderLink(this.buildQuerySearch(active + 1), Paginator.Titles.Next) }
                { this.renderLink(this.buildQuerySearch(total), Paginator.Titles.Last) }
            </div>
        )
    }
}

export default Paginator;
