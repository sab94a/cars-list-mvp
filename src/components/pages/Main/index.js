//@flow

import React from 'react';
import qs from 'query-string';

import type { PagesNavigation } from 'types/routes';
import type { CarView, SelectView } from 'types/views';
import type { CarsFilters } from 'types/api';
import type { State as FilterFormState } from 'components/pages/Main/FiltersForm';

import { CARS_PER_PAGE } from 'constants/api';

import Card from 'components/lib/Card';
import Link from 'components/lib/Link';
import Select from 'components/lib/Select';
import Paginator from 'components/lib/Paginator';
import FiltersForm from 'components/pages/Main/FiltersForm';

import styles from './index.module.scss';

export type Props = {
    init: (params: mixed) => void,
    update: (params: mixed) => void,
    removeFavourite: (id: number) => void,
    navigate: (params: mixed) => void,
    cars: Array<CarView>,
    colors: Array<SelectView>,
    sortings: Array<SelectView>,
    manufacturers: Array<SelectView>,
    carsLoading: boolean,
    navigation: ?PagesNavigation,
    filters: CarsFilters,
    search: string,
}

class Main extends React.PureComponent<Props> {
    static SortLabel = 'Sort by';

    handlers = {};

    placeholder:Array<boolean> = new Array(CARS_PER_PAGE).fill(false);

    componentDidMount() {
        this.props.init(this.getParams())
    };

    componentDidUpdate({ search: prevSearch }: Props) {
        const { search, update } = this.props;

        if (search !== prevSearch) {
            update(this.getParams())
        }
    }

    getParams() {
        const { search } = this.props;
        const { page, manufacturer, color, sort } = qs.parse(search);

        return {
            page: page ? Number.parseInt(page) : 1, 
            manufacturer,
            color,
            sort
        };
    }

    onFilterChange = (params: CarsFilters) => {
        const { navigate } = this.props

        let query = {}

        for(let key in params) {
            if(params[key]) {
                query[key] = params[key]
            }
        }

        query.page = 1

        navigate({
            search: `?${qs.stringify(query)}`
        })
    }

    onFilterFormSubmit = ({ manufacturer, color}: FilterFormState) => {
        this.onFilterChange({
            ...this.getParams(),
            manufacturer,
            color
        })
    }

    onShortChange = (sort: ?string) => {
        this.onFilterChange({
            ...this.getParams(),
            sort
        })
    };

    bindRemoveClick = (carId: number) => {
        const { removeFavourite } = this.props;

        if(!this.handlers[carId]) {
            this.handlers[carId] = e => {
                e.preventDefault()

                removeFavourite(carId)
            }
        }
        return this.handlers[carId];
    }

    renderCarFooter(car: CarView) {
        return (
             <>
                <Link href={ car.link }>View details</Link>
                { car.isFavourite && (
                    <>
                        <span className={ styles.separator } />
                        <Link href="#" onClick={ this.bindRemoveClick(car.stockNumber) }>
                            Remove from favourites
                        </Link>
                    </>
                ) }
            </>
        )
    }

    render() {
        const { 
            cars,
            colors,
            sortings,
            manufacturers,
            navigation,
            carsLoading,
            filters
        } = this.props;

        const isNotFound = !carsLoading && !cars.length;
        const shownItems = navigation ? navigation.shownItems : cars.length;
        const totalItems = navigation ? navigation.totalItems : cars.length;

        return (
            <div className={ styles.root }>
                <aside className={ styles.aside }>
                    <FiltersForm 
                        color={ filters.color }
                        colors={ colors }
                        manufacturer={ filters.manufacturer }
                        manufacturers={ manufacturers }
                        onSubmit={ this.onFilterFormSubmit }
                    />
                </aside>
                <div className={ styles.content }>
                    <div className={ styles.searchHeader}>
                        <div className={ styles.searchInfo}>
                            <h2>Available cars</h2>
                            <div className={ styles.searchResult}>
                                { `Showing ${ shownItems } of ${ totalItems } results` }
                            </div>
                        </div>
                        <div className={ styles.sort }>
                            <Select 
                                label={ Main.SortLabel }
                                value={ filters.sort }
                                options={ sortings }
                                onChange={ this.onShortChange }
                            />
                        </div>
                    </div>
                    { !isNotFound && (
                        <ul className={ styles.items }>
                            { !carsLoading && cars.map((item, index) => (
                                <Card 
                                    key={ index }
                                    elemType='li'
                                    title={ item.title }
                                    image={ item.image }
                                    description={ item.description }
                                    footer={this.renderCarFooter(item)}
                                />
                            ))}

                            { carsLoading && this.placeholder.map((item, index) => (
                                <Card key={ index } elemType='li' loading={ true } />
                            )) }
                        </ul>
                    ) }
                    
                    { isNotFound && (
                        <div className={ styles.notFound }>There is no avalible cars for your search.</div>
                    ) }

                    { !!navigation && (
                        <Paginator 
                            active={ navigation.page } 
                            total={ navigation.totalPages } 
                        />
                    ) }
                </div>
            </div>
        )
    };
}

export default Main;
