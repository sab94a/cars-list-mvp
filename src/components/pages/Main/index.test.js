import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import Card from 'components/lib/Card';
import Link from 'components/lib/Link';
import Paginator from 'components/lib/Paginator';
import Main from './';


describe('<Main /> page component', () => {
    let fetchData;
    const cars = new Array(10).fill(10).map((item, index) => ({
        stockNumber: 1,
        title: `title ${index}`,
        image: `image_${index}.png`,
        link: `?cars/${index}/`,
        description: `description_${index}`
    }));

    const navigation = {
        shownItems: 10,
        totalItems: 1000,
        totalPages: 100,
        page: 10
    };
    const location = {
        search: ''
    };

    let component;

    const create = params => {
        component = mount(<Main { ...params } />);
    };

    beforeEach(() => {
        fetchData = jest.fn();
    })

    it('Should render correctly', () => {
        create({ fetchData, cars, navigation, location, carsLoading: false });

        expect(toJson(component)).toMatchSnapshot();
    });

    it('Should show correct search result', () => {
        const $search = component.find('.searchResult');

        expect($search.text()).toBe(`Showing ${ navigation.shownItems } of ${ navigation.totalItems } results`)
    });

    it('Should show render list correctly', () => {
        const $cards = component.find(Card)

        expect($cards).toHaveLength(cars.length)

        cars.forEach((item, index) => {
            const $card = $cards.at(index);

            expect($card.prop('title')).toBe(item.title);
            expect($card.prop('image')).toBe(item.image);
            expect($card.prop('description')).toBe(item.description);
            expect($card.contains(<Link href={ item.link }>View details</Link>)).toEqual(true);
        });
    });

    it('If loading, should render loading placeholder', () => {
        create({ fetchData, navigation, location, carsLoading: true });

        const $cards = component.find(Card);

        expect($cards).toHaveLength(cars.length);

        $cards.forEach((item, index) => {
            expect($cards.at(index).prop('loading')).toBe(true);
        });
    });

    it('Should render Paginator if component has pages', () => {
        const $paginator = component.find(Paginator);

        expect($paginator).toHaveLength(1);
        expect($paginator.prop('active')).toBe(navigation.page);
        expect($paginator.prop('total')).toBe(navigation.totalPages);
    });

    it('Shouldn\'t render paginator if pages are not passed', () => {
        create({ fetchData, navigation: { totalPages: 0 }, location, carsLoading: true });

        expect(component.find(Paginator)).toHaveLength(0);
    });

    it('Should render Not Found Node insted of cars correctly', () => {
        create({ fetchData, navigation: {}, cars: [], location, carsLoading: false });

        expect(component.find('.notFound')).toHaveLength(1);
    })
});
