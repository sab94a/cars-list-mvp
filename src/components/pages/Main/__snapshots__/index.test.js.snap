// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Main /> page component Should render correctly 1`] = `
<Main
  cars={
    Array [
      Object {
        "description": "description_0",
        "image": "image_0.png",
        "link": "?cars/0/",
        "stockNumber": 1,
        "title": "title 0",
      },
      Object {
        "description": "description_1",
        "image": "image_1.png",
        "link": "?cars/1/",
        "stockNumber": 1,
        "title": "title 1",
      },
      Object {
        "description": "description_2",
        "image": "image_2.png",
        "link": "?cars/2/",
        "stockNumber": 1,
        "title": "title 2",
      },
      Object {
        "description": "description_3",
        "image": "image_3.png",
        "link": "?cars/3/",
        "stockNumber": 1,
        "title": "title 3",
      },
      Object {
        "description": "description_4",
        "image": "image_4.png",
        "link": "?cars/4/",
        "stockNumber": 1,
        "title": "title 4",
      },
      Object {
        "description": "description_5",
        "image": "image_5.png",
        "link": "?cars/5/",
        "stockNumber": 1,
        "title": "title 5",
      },
      Object {
        "description": "description_6",
        "image": "image_6.png",
        "link": "?cars/6/",
        "stockNumber": 1,
        "title": "title 6",
      },
      Object {
        "description": "description_7",
        "image": "image_7.png",
        "link": "?cars/7/",
        "stockNumber": 1,
        "title": "title 7",
      },
      Object {
        "description": "description_8",
        "image": "image_8.png",
        "link": "?cars/8/",
        "stockNumber": 1,
        "title": "title 8",
      },
      Object {
        "description": "description_9",
        "image": "image_9.png",
        "link": "?cars/9/",
        "stockNumber": 1,
        "title": "title 9",
      },
    ]
  }
  carsLoading={false}
  fetchData={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "color": undefined,
            "manufacturer": undefined,
            "page": 1,
            "sort": undefined,
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  location={
    Object {
      "search": "",
    }
  }
  navigation={
    Object {
      "page": 10,
      "shownItems": 10,
      "totalItems": 1000,
      "totalPages": 100,
    }
  }
>
  <div
    className="root"
  >
    <aside
      className="aside"
    >
      Heading Asside
    </aside>
    <div
      className="content"
    >
      <div
        className="searchResult"
      >
        Showing 10 of 1000 results
      </div>
      <ul
        className="items"
      >
        <Memo(Card)
          description="description_0"
          elemType="li"
          footer={
            <Link
              href="?cars/0/"
            >
              View details
            </Link>
          }
          image="image_0.png"
          key="0"
          title="title 0"
        >
          <li
            className="root"
          >
            <div
              className="image"
              style={
                Object {
                  "backgroundImage": "url(image_0.png)",
                }
              }
            />
            <div
              className="content"
            >
              <h3
                className="title"
              >
                title 0
              </h3>
              <p
                className="description"
              >
                description_0
              </p>
              <div
                className="footer"
              >
                <Link
                  href="?cars/0/"
                >
                  <a
                    href="?cars/0/"
                  >
                    View details
                  </a>
                </Link>
              </div>
            </div>
          </li>
        </Memo(Card)>
        <Memo(Card)
          description="description_1"
          elemType="li"
          footer={
            <Link
              href="?cars/1/"
            >
              View details
            </Link>
          }
          image="image_1.png"
          key="1"
          title="title 1"
        >
          <li
            className="root"
          >
            <div
              className="image"
              style={
                Object {
                  "backgroundImage": "url(image_1.png)",
                }
              }
            />
            <div
              className="content"
            >
              <h3
                className="title"
              >
                title 1
              </h3>
              <p
                className="description"
              >
                description_1
              </p>
              <div
                className="footer"
              >
                <Link
                  href="?cars/1/"
                >
                  <a
                    href="?cars/1/"
                  >
                    View details
                  </a>
                </Link>
              </div>
            </div>
          </li>
        </Memo(Card)>
        <Memo(Card)
          description="description_2"
          elemType="li"
          footer={
            <Link
              href="?cars/2/"
            >
              View details
            </Link>
          }
          image="image_2.png"
          key="2"
          title="title 2"
        >
          <li
            className="root"
          >
            <div
              className="image"
              style={
                Object {
                  "backgroundImage": "url(image_2.png)",
                }
              }
            />
            <div
              className="content"
            >
              <h3
                className="title"
              >
                title 2
              </h3>
              <p
                className="description"
              >
                description_2
              </p>
              <div
                className="footer"
              >
                <Link
                  href="?cars/2/"
                >
                  <a
                    href="?cars/2/"
                  >
                    View details
                  </a>
                </Link>
              </div>
            </div>
          </li>
        </Memo(Card)>
        <Memo(Card)
          description="description_3"
          elemType="li"
          footer={
            <Link
              href="?cars/3/"
            >
              View details
            </Link>
          }
          image="image_3.png"
          key="3"
          title="title 3"
        >
          <li
            className="root"
          >
            <div
              className="image"
              style={
                Object {
                  "backgroundImage": "url(image_3.png)",
                }
              }
            />
            <div
              className="content"
            >
              <h3
                className="title"
              >
                title 3
              </h3>
              <p
                className="description"
              >
                description_3
              </p>
              <div
                className="footer"
              >
                <Link
                  href="?cars/3/"
                >
                  <a
                    href="?cars/3/"
                  >
                    View details
                  </a>
                </Link>
              </div>
            </div>
          </li>
        </Memo(Card)>
        <Memo(Card)
          description="description_4"
          elemType="li"
          footer={
            <Link
              href="?cars/4/"
            >
              View details
            </Link>
          }
          image="image_4.png"
          key="4"
          title="title 4"
        >
          <li
            className="root"
          >
            <div
              className="image"
              style={
                Object {
                  "backgroundImage": "url(image_4.png)",
                }
              }
            />
            <div
              className="content"
            >
              <h3
                className="title"
              >
                title 4
              </h3>
              <p
                className="description"
              >
                description_4
              </p>
              <div
                className="footer"
              >
                <Link
                  href="?cars/4/"
                >
                  <a
                    href="?cars/4/"
                  >
                    View details
                  </a>
                </Link>
              </div>
            </div>
          </li>
        </Memo(Card)>
        <Memo(Card)
          description="description_5"
          elemType="li"
          footer={
            <Link
              href="?cars/5/"
            >
              View details
            </Link>
          }
          image="image_5.png"
          key="5"
          title="title 5"
        >
          <li
            className="root"
          >
            <div
              className="image"
              style={
                Object {
                  "backgroundImage": "url(image_5.png)",
                }
              }
            />
            <div
              className="content"
            >
              <h3
                className="title"
              >
                title 5
              </h3>
              <p
                className="description"
              >
                description_5
              </p>
              <div
                className="footer"
              >
                <Link
                  href="?cars/5/"
                >
                  <a
                    href="?cars/5/"
                  >
                    View details
                  </a>
                </Link>
              </div>
            </div>
          </li>
        </Memo(Card)>
        <Memo(Card)
          description="description_6"
          elemType="li"
          footer={
            <Link
              href="?cars/6/"
            >
              View details
            </Link>
          }
          image="image_6.png"
          key="6"
          title="title 6"
        >
          <li
            className="root"
          >
            <div
              className="image"
              style={
                Object {
                  "backgroundImage": "url(image_6.png)",
                }
              }
            />
            <div
              className="content"
            >
              <h3
                className="title"
              >
                title 6
              </h3>
              <p
                className="description"
              >
                description_6
              </p>
              <div
                className="footer"
              >
                <Link
                  href="?cars/6/"
                >
                  <a
                    href="?cars/6/"
                  >
                    View details
                  </a>
                </Link>
              </div>
            </div>
          </li>
        </Memo(Card)>
        <Memo(Card)
          description="description_7"
          elemType="li"
          footer={
            <Link
              href="?cars/7/"
            >
              View details
            </Link>
          }
          image="image_7.png"
          key="7"
          title="title 7"
        >
          <li
            className="root"
          >
            <div
              className="image"
              style={
                Object {
                  "backgroundImage": "url(image_7.png)",
                }
              }
            />
            <div
              className="content"
            >
              <h3
                className="title"
              >
                title 7
              </h3>
              <p
                className="description"
              >
                description_7
              </p>
              <div
                className="footer"
              >
                <Link
                  href="?cars/7/"
                >
                  <a
                    href="?cars/7/"
                  >
                    View details
                  </a>
                </Link>
              </div>
            </div>
          </li>
        </Memo(Card)>
        <Memo(Card)
          description="description_8"
          elemType="li"
          footer={
            <Link
              href="?cars/8/"
            >
              View details
            </Link>
          }
          image="image_8.png"
          key="8"
          title="title 8"
        >
          <li
            className="root"
          >
            <div
              className="image"
              style={
                Object {
                  "backgroundImage": "url(image_8.png)",
                }
              }
            />
            <div
              className="content"
            >
              <h3
                className="title"
              >
                title 8
              </h3>
              <p
                className="description"
              >
                description_8
              </p>
              <div
                className="footer"
              >
                <Link
                  href="?cars/8/"
                >
                  <a
                    href="?cars/8/"
                  >
                    View details
                  </a>
                </Link>
              </div>
            </div>
          </li>
        </Memo(Card)>
        <Memo(Card)
          description="description_9"
          elemType="li"
          footer={
            <Link
              href="?cars/9/"
            >
              View details
            </Link>
          }
          image="image_9.png"
          key="9"
          title="title 9"
        >
          <li
            className="root"
          >
            <div
              className="image"
              style={
                Object {
                  "backgroundImage": "url(image_9.png)",
                }
              }
            />
            <div
              className="content"
            >
              <h3
                className="title"
              >
                title 9
              </h3>
              <p
                className="description"
              >
                description_9
              </p>
              <div
                className="footer"
              >
                <Link
                  href="?cars/9/"
                >
                  <a
                    href="?cars/9/"
                  >
                    View details
                  </a>
                </Link>
              </div>
            </div>
          </li>
        </Memo(Card)>
      </ul>
      <Paginator
        active={10}
        total={100}
      >
        <div
          className="root"
        >
          <Link
            className="link"
            href="?page=1"
          >
            <a
              className="link"
              href="?page=1"
            >
              First
            </a>
          </Link>
          <Link
            className="link"
            href="?page=9"
          >
            <a
              className="link"
              href="?page=9"
            >
              Previous
            </a>
          </Link>
          <div
            className="content"
          >
            Page 10 of 100
          </div>
          <Link
            className="link"
            href="?page=11"
          >
            <a
              className="link"
              href="?page=11"
            >
              Next
            </a>
          </Link>
          <Link
            className="link"
            href="?page=100"
          >
            <a
              className="link"
              href="?page=100"
            >
              Last
            </a>
          </Link>
        </div>
      </Paginator>
    </div>
  </div>
</Main>
`;
